src/providers/ad/ad_pac_common.c

```diff
@@ -67,17 +244,78 @@ errno_t ad_get_data_from_pac(TALLOC_CTX *mem_ctx,
         goto done;
     }
 
+    if (_logon_info != NULL) {
+        *_logon_info = NULL;
+    }
+    if (_upn_dns_info != NULL) {
+        *_upn_dns_info = NULL;
+    }
+    if (_attributes_info != NULL) {
+        *_attributes_info = NULL;
+    }
+    if (_requester_sid != NULL) {
+        *_requester_sid = NULL;
+    }
+
     for(c = 0; c < pac_data->num_buffers; c++) {
-        if (pac_data->buffers[c].type == PAC_TYPE_LOGON_INFO) {
-            *_logon_info = talloc_steal(mem_ctx,
+        switch (pac_data->buffers[c].type) {
+        case PAC_TYPE_SRV_CHECKSUM:
+            break;
+        case PAC_TYPE_KDC_CHECKSUM:
+            break;
+        case PAC_TYPE_LOGON_INFO:
+            if (_logon_info != NULL) {
+                *_logon_info = talloc_steal(mem_ctx,
                                     pac_data->buffers[c].info->logon_info.info);
+            }
+            break;
+        case PAC_TYPE_UPN_DNS_INFO:
+            if (_upn_dns_info != NULL) {
+                *_upn_dns_info = talloc_steal(mem_ctx,
+                                      &pac_data->buffers[c].info->upn_dns_info);
+            }
+            break;
+#ifdef HAVE_STRUCT_PAC_ATTRIBUTES_INFO_FLAGS
+        case PAC_TYPE_ATTRIBUTES_INFO:
+            if (_attributes_info != NULL) {
+                *_attributes_info = talloc_steal(mem_ctx,
+                                   &pac_data->buffers[c].info->attributes_info);
+            }
+            break;
+#endif
+#ifdef HAVE_STRUCT_PAC_REQUESTER_SID_SID
+        case PAC_TYPE_REQUESTER_SID:
+            if (_requester_sid != NULL) {
+                *_requester_sid = talloc_steal(mem_ctx,
+                                     &pac_data->buffers[c].info->requester_sid);
+            }
+            break;
+#endif
+        default:
+            DEBUG(SSSDBG_TRACE_ALL, "Unhandled PAC buffer type [%d].\n",
+                                    pac_data->buffers[c].type);
+        }
+    }
+
+    if (_logon_info != NULL && *_logon_info == NULL) {
+        ret = ERR_CHECK_PAC_FAILED;
+        goto done;
+    }
 
-            ret = EOK;
+    /* Make sure the content of different PAC buffers is consistent. */
+    if (_logon_info != NULL && _upn_dns_info != NULL) {
+        ret = check_logon_info_upn_dns_info(*_logon_info, *_upn_dns_info);
+        if (ret != EOK) {
+            DEBUG(SSSDBG_OP_FAILURE, "Validating PAC data failed.\n");
+            talloc_free(*_logon_info);
+            *_logon_info = NULL;
+            talloc_free(*_upn_dns_info);
+            *_upn_dns_info = NULL;
```

Comment:
`_attributes_info` and `_requester_sid`?

Reply:
I removed both because those buffers should be only present in a TGT and we can only look into service tickets.